### PR TITLE
Add DSL for creating Goals

### PR DESCRIPTION
**Pull request by cdupuis**
This allows to create `Goals` instances without the tedious deconstruction etc:

```
    const StagingDeploymentGoals = goals("Simple Goals")
            .plan(ReviewGoal)
            .plan(BuildGoal, AutofixGoal).after(ReviewGoal)
            .plan(StagingEndpointGoal).after(BuildGoal)
            .plan(ProductionDeploymentGoal).after(BuildGoal, StagingEndpointGoal);
```

### CODE REVIEW

File: test/api/goal/GoalsTest.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+} from "../../../src";
+
+describe("GoalBuilder", () => {
+
+    it("should construct simple goal set", () => {
+        const simpleGoals = goals("Simple Goals")
+            .plan(BuildGoal, AutofixGoal)
+            .plan(StagingEndpointGoal);
+
+        assert.equal(simpleGoals.name, "Simple Goals");
```
[the remark this answers is not in the document]
Comment (cdupuis) [reply test/api/goal/GoalsTest.ts]: I find the error messages are just nicer if you use `assert.equals` compared to `assert(a === b)`. That way you get the actual difference in the test. I'll try `strictEquals`.

File: test/api/goal/GoalsTest.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            .plan(ReviewGoal)
+            .plan(BuildGoal, AutofixGoal).after(ReviewGoal)
+            .plan(StagingEndpointGoal).after(BuildGoal)
+            .plan(ProductionDeploymentGoal).after(BuildGoal, StagingEndpointGoal);
+
+        assert.equal(simpleGoals.name, "Simple Goals");
+        assert.equal(simpleGoals.goals.length, 5);
+
+        const buildGoal = simpleGoals.goals.find(g => g.name === BuildGoal.name) as GoalWithPrecondition;
+        assert.equal(buildGoal.name, buildGoal.name);
```
[the remark this answers is not in the document]
Comment (cdupuis) [reply test/api/goal/GoalsTest.ts]: Opps.

File: src/api/goal/Goals.ts [hunk 3]
```diff
@@ -31,3 +34,98 @@ export class Goals {
...
+    public after(...newGoals: Array<Goal>): Goals & GoalsBuilder {
+        const lastGoalsWithPreConditions = [];
+
+        // Add the preCondition into the last added goals
+        this.lastGoals.forEach(g => {
+            lastGoalsWithPreConditions.push(new GoalWithPrecondition(g.definition, ...newGoals));
+        });
+
+        // Remove the previously added goals
+        this.lastGoals.forEach(g => {
```
[the remark this answers is not in the document]
Comment (cdupuis) [reply src/api/goal/Goals.ts]: Yes. good point.

File: src/api/goal/Goals.ts [hunk 4]
```diff
@@ -31,3 +34,98 @@ export class Goals {
...
+
+}
+
+/**
+ * Extension to GoalsBuilder allowing to add preConditions,
+ */
+export interface GoalsAndPreConditionBuilder extends GoalsBuilder {
+
+    /**
+     * Add a preCondition to previously added goal or goals
```
[the remark this answers is not in the document]
Comment (cdupuis) [reply src/api/goal/Goals.ts]: Exactly. It just adds the given Goals as preCondition.

**Comment by cdupuis** [comment]
Thanks for the review @ddgenome. I addressed your comments (hopefully).

File: test/api/goal/GoalsTest.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddgenome) [reply test/api/goal/GoalsTest.ts]: String, I get the unified diff output even with `===`.

File: src/api/goal/Goals.ts [hunk 5]
```diff
@@ -31,3 +34,95 @@ export class Goals {
...
+    }
+
+    public after(...newGoals: Array<Goal>): Goals & GoalsBuilder {
+        const lastGoalsWithPreConditions = [];
+
+        this.lastGoals.forEach(g => {
+            // Add the preCondition into the last added goals
+            lastGoalsWithPreConditions.push(new GoalWithPrecondition(g.definition, ...newGoals));
+            // Remove the previously added goals
+            this.goals.splice(this.goals.indexOf(g), 1);
```
Comment (ddgenome) [inline src/api/goal/Goals.ts]: If the first argument to splice is a negative number, it splices counting from the end. So if the goal is not present for some reason, bad things would happen.

Comment (cdupuis) [reply src/api/goal/Goals.ts]: yeah, I'll make it an error.

File: src/api/goal/Goals.ts [hunk 6]
```diff
@@ -31,3 +34,100 @@ export class Goals {
...
+
+/**
+ * Create Goals instance using a fluent API.
+ *
+ *  const simpleGoals = goals("Simple Goals")
+ *     .plan(ReviewGoal)
+ *     .plan(BuildGoal, AutofixGoal).after(ReviewGoal)
+ *     .plan(StagingEndpointGoal).after(BuildGoal)
+ *     .plan(ProductionDeploymentGoal).after(BuildGoal, StagingEndpointGoal);
+ *
```
Comment (johnsonr) [inline src/api/goal/Goals.ts]: This could probably use some Markdown love to get formatted properly in Typedoc. We should consider the formatting.

**Pull request merged**